commit c5faa351c1ef6d4555478a7f50b3a16ece7e0b2a
Author: jim-p <jimp@netgate.com>
Date:   Tue Mar 28 09:09:24 2023 -0400

    Message queue handling improvements. Fixes #14031 Fixes #14061
    
    * Do not attempt to process the message queue without sufficient
      privileges.
    * Check for permission to write the message queue and for write
      failures and bail early if there is a problem.

diff --git a/src/etc/inc/notices.inc b/src/etc/inc/notices.inc
index 539698b821..5a9f4f617d 100644
--- a/src/etc/inc/notices.inc
+++ b/src/etc/inc/notices.inc
@@ -246,16 +246,26 @@ function notices_sendqueue() {
 	$messagequeue = array();
 
 	while(true) {
-		$notifyqueue_lck = lock("notifyqueue", LOCK_EX);
 		$nothing_done_count++;
 		$smtpcount = 0;
 		$messages = array();
-		if (file_exists("{$g['vardb_path']}/notifyqueue.messages")) {
+		if (file_exists("{$g['vardb_path']}/notifyqueue.messages") &&
+		    is_writable("{$g['vardb_path']}/notifyqueue.messages")) {
+			$notifyqueue_lck = lock("notifyqueue", LOCK_EX);
 			$messages = unserialize(file_get_contents("{$g['vardb_path']}/notifyqueue.messages"));
 			$messagequeue = $messages;
 			$messages['mails']['item'] = array(); // clear all items to be send
-			file_put_contents("{$g['vardb_path']}/notifyqueue.messages", serialize($messages));
+			$ret = file_put_contents("{$g['vardb_path']}/notifyqueue.messages", serialize($messages));
+			if ($ret === false) {
+				log_error("ERROR: Failed to write notify message queue!");
+				return;
+			}
 			unset($messages);
+		} else {
+			/* Queue does not exist or is not writable, so no action can be taken
+			 * https://redmine.pfsense.org/issues/14031
+			 */
+			return;
 		}
 		// clear lock before trying to send messages, so new one's can be added
 		unlock($notifyqueue_lck);
@@ -296,11 +306,21 @@ function notify_via_queue_add($message, $type='mail') {
 	$notifyqueue_lck = lock("notifyqueue", LOCK_EX);
 	$messages = array();
 	if (file_exists("{$g['vardb_path']}/notifyqueue.messages")) {
+		if (!is_writable("{$g['vardb_path']}/notifyqueue.messages")) {
+			/* Cannot write to notify queue, so exit early
+			 * https://redmine.pfsense.org/issues/14031
+			 */
+			return;
+		}
 		$messages = unserialize(file_get_contents("{$g['vardb_path']}/notifyqueue.messages"));
 	}
-	if(is_array($messages)) {
+	if (is_array($messages)) {
 		$messages['mails']['item'][] = $mail;
-		file_put_contents("{$g['vardb_path']}/notifyqueue.messages", serialize($messages));
+		$ret = file_put_contents("{$g['vardb_path']}/notifyqueue.messages", serialize($messages));
+		if ($ret === false) {
+			log_error("ERROR: Failed to write notify message queue!");
+			return;
+		}
 	}
 	unset($messages);
 
diff --git a/src/usr/local/bin/notify_monitor.php b/src/usr/local/bin/notify_monitor.php
index 1b70087f39..b6b7f5a7ef 100755
--- a/src/usr/local/bin/notify_monitor.php
+++ b/src/usr/local/bin/notify_monitor.php
@@ -20,6 +20,18 @@
  * limitations under the License.
  */
 
+/* Do not attempt to take any action if this script is invoked by an
+ * unprivileged user. Such users are unable to write to some files such as
+ * the config.cache and the message queue which can lead to PHP errors and/or
+ * duplicate notifications.
+ * https://redmine.pfsense.org/issues/14031
+ * https://redmine.pfsense.org/issues/14061
+ */
+if (posix_geteuid() !== 0) {
+	echo "This script cannot be run by an unprivileged user.\n";
+	exit(1);
+}
+
 include_once('util.inc');
 include_once('notices.inc');
 
